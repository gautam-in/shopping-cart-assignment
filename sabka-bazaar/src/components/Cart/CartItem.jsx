import PropTypes from "prop-types";
import { Button } from "../Button/Button";

export const CartItem = ({
  name,
  imageURL,
  price,
  quantity,
  id,
  incrementQuantity,
  decrementQuantity,
}) => {
  return (
    <div className="cart-item-container">
      <div className="cart-item-img-container">
        <img
          className="cart-item-img"
          src={imageURL}
          alt={name}
          width="60"
          height="60"
        />
      </div>
      <div className="cart-item-info">
        <div className="cart-item-desc">
          <div className="cart-item-name">{name}</div>
          <div>
            <Button
              title="-"
              type="circularBtn"
              onClick={() => decrementQuantity(id)}
            />
            <span className="quantity">{quantity}</span>
            <Button
              title="+"
              type="circularBtn"
              onClick={() => incrementQuantity(id)}
            />
            <span className="quantity">x</span> Rs.{price}
          </div>
        </div>

        <div className="cart-item-total">Rs.{quantity * price}</div>
      </div>
    </div>
  );
};
CartItem.propTypes = {
  name: PropTypes.string,
  imageURL: PropTypes.string,
  price: PropTypes.number,
  quantity: PropTypes.number,
  id: PropTypes.string,
  incrementQuantity: PropTypes.func,
  decrementQuantity: PropTypes.func,
};
