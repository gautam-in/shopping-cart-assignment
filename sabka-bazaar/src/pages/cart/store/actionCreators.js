import { ADD_TO_CART } from "./types";

export const addToCart = (product) => {
  return {
    type: ADD_TO_CART,
    payload: product,
  };
};
