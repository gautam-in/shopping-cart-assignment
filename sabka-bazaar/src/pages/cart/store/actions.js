import { ADD_TO_CART, UPDATE_CART } from "./types";
import { createAction } from "@reduxjs/toolkit";

export const addToCart = createAction(ADD_TO_CART);

export const updateCartItems = createAction(UPDATE_CART);
