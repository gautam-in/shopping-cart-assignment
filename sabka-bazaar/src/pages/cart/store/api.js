export const postCartItemToServer = async (options) => {
  try {
    const response = await fetch(
      `${process.env.REACT_APP_BASEURL}/addToCart`,
      options
    );
    return response.json();
  } catch (error) {
    throw new Error(error.message);
  }
};
