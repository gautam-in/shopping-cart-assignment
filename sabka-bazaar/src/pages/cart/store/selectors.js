import { createSelector } from "@reduxjs/toolkit";

const selectSlice = (state) => state["cart"] || {};

export const selectIsAddingToCartLoading = createSelector(
  [selectSlice],
  (state) => state.addToCartLoading
);

export const selectCartItems = createSelector(
  [selectSlice],
  (state) => state.cart
);

export const selectCartItemsCount = createSelector(
  [selectCartItems],
  (cartItems) =>
    cartItems.reduce(
      (cartItemsCount, cartItem) => cartItemsCount + cartItem.quantity,
      0
    )
);

export const selectCartTotal = createSelector([selectCartItems], (cartItems) =>
  cartItems.reduce(
    (cartTotal, cartItem) => cartTotal + cartItem.quantity * cartItem.price,
    0
  )
);
