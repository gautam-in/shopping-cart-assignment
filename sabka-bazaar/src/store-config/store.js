import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";
import homeReducer from "../pages/home/store/slice";
import productsReducer from "../pages/products/store/slice";
import cartReducer from "../pages/cart/store/slice";
import rootSaga from "./rootSaga";
const sagaMiddleware = createSagaMiddleware();
export const store = configureStore({
  reducer: {
    home: homeReducer,
    products: productsReducer,
    cart: cartReducer,
  },
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);
