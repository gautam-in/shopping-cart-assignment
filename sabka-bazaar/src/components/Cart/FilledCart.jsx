import PropTypes from "prop-types";
import { useDispatch, useSelector } from "react-redux";
import { updateCartItems } from "../../pages/cart/store/actions";
import {
  selectCartItems,
  selectCartTotal,
} from "../../pages/cart/store/selectors";
import { Button } from "../Button/Button";
import { CartItem } from "./CartItem";
import "./cart.styles.scss";

export const FilledCart = ({ handleClose }) => {
  const cartItems = useSelector(selectCartItems);
  const cartTotal = useSelector(selectCartTotal);

  const dispatch = useDispatch();
  const incrementQuantity = (productId) => {
    const updatedCartItems = cartItems.map((cartItem) =>
      cartItem.id === productId
        ? {
            ...cartItem,
            quantity: cartItem.quantity + 1,
          }
        : cartItem
    );
    dispatch(updateCartItems(updatedCartItems));
  };
  const decrementQuantity = (productId) => {
    const currentCartItem = cartItems.find(
      (cartItem) => cartItem.id === productId
    );
    let updatedCartItems = [];
    if (currentCartItem.quantity > 1) {
      updatedCartItems = cartItems.map((cartItem) =>
        cartItem.id === productId
          ? {
              ...cartItem,
              quantity: cartItem.quantity - 1,
            }
          : cartItem
      );
    } else {
      updatedCartItems = cartItems.filter(
        (cartItem) => cartItem.id !== productId
      );
    }
    dispatch(updateCartItems(updatedCartItems));
  };
  return (
    <>
      <div className="cart-items-list">
        {cartItems.map(({ name, id, imageURL, price, quantity }) => {
          return (
            <CartItem
              key={id}
              name={name}
              imageURL={imageURL}
              price={price}
              id={id}
              quantity={quantity}
              incrementQuantity={incrementQuantity}
              decrementQuantity={decrementQuantity}
            />
          );
        })}
      </div>
      <div className="price-tag-container">
        <div>
          <img
            src="/static/images/lowest-price.png"
            alt="lowest price guaranteed image"
            width="60"
            height="30"
            className="lowest-price-tag"
          />
        </div>
        <div>
          <p>You won&apos;t find it cheaper anywhere</p>
        </div>
      </div>
      <div className="checkout-section">
        <p>Promo code can be applied on payment page</p>
        <div className="checkout-btn-container">
          <Button
            title="Proceed to Checkout"
            type="checkoutBtn"
            price={cartTotal}
            onClick={handleClose}
          />
        </div>
      </div>
    </>
  );
};

FilledCart.propTypes = {
  cartItems: PropTypes.array,
  handleClose: PropTypes.func,
};
