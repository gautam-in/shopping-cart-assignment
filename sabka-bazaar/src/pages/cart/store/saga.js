import { ADD_TO_CART, UPDATE_CART } from "./types";
import {
  addToCartLoading,
  addToCartSuccess,
  addToCartError,
  updateCart,
} from "./slice";
import { call, put, takeLatest } from "redux-saga/effects";
import { postCartItemToServer } from "./api";

const addToCartHelper = ({ cartItems, productToAdd }) => {
  const existingCartItem = cartItems.find((cartItem) => {
    return cartItem.id === productToAdd.id;
  });
  if (existingCartItem) {
    return cartItems.map((cartItem) =>
      cartItem.id === productToAdd.id
        ? { ...cartItem, quantity: cartItem.quantity + 1 }
        : cartItem
    );
  }

  return [...cartItems, { ...productToAdd, quantity: 1 }];
};

function* handleCartData({ payload }) {
  const options = {
    method: "POST",
    data: { id: payload.productToAdd.id },
  };

  try {
    const response = yield call(postCartItemToServer, options);
    if (response) {
      const cartItems = addToCartHelper(payload);
      yield put(addToCartSuccess(cartItems));
    }
  } catch (error) {
    yield put(addToCartError(error.message));
  } finally {
    yield put(addToCartLoading(false));
  }
}

function* updateCartHandler({ payload }) {
  yield put(updateCart(payload));
}

function* cartSaga() {
  yield takeLatest(ADD_TO_CART, handleCartData);
  yield takeLatest(UPDATE_CART, updateCartHandler);
}

export default cartSaga;
