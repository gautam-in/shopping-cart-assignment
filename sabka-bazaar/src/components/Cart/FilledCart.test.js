import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../store-config/store";
import { CartItem } from "./CartItem";
import { FilledCart } from "./FilledCart";
let container;
beforeEach(() => {
  const mockProps = {
    handleClose: jest.fn(),
  };
  container = render(
    <Provider store={store}>
      <FilledCart props={mockProps} />
    </Provider>
  );
});
describe("Filled cart component", () => {
  it("should match snapshot", () => {
    expect(container).toMatchSnapshot();
  });
});
