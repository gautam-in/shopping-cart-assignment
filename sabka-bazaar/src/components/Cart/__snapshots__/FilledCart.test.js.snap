// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filled cart component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cart-items-list"
      />
      <div
        class="price-tag-container"
      >
        <div>
          <img
            alt="lowest price guaranteed image"
            class="lowest-price-tag"
            height="30"
            src="/static/images/lowest-price.png"
            width="60"
          />
        </div>
        <div>
          <p>
            You won't find it cheaper anywhere
          </p>
        </div>
      </div>
      <div
        class="checkout-section"
      >
        <p>
          Promo code can be applied on payment page
        </p>
        <div
          class="checkout-btn-container"
        >
          <button
            class="checkout-btn"
          >
            <div>
              Proceed to Checkout
            </div>
            <div
              class="price-container"
            >
              <div>
                Rs.
                0
              </div>
              <div>
                &gt;
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cart-items-list"
    />
    <div
      class="price-tag-container"
    >
      <div>
        <img
          alt="lowest price guaranteed image"
          class="lowest-price-tag"
          height="30"
          src="/static/images/lowest-price.png"
          width="60"
        />
      </div>
      <div>
        <p>
          You won't find it cheaper anywhere
        </p>
      </div>
    </div>
    <div
      class="checkout-section"
    >
      <p>
        Promo code can be applied on payment page
      </p>
      <div
        class="checkout-btn-container"
      >
        <button
          class="checkout-btn"
        >
          <div>
            Proceed to Checkout
          </div>
          <div
            class="price-container"
          >
            <div>
              Rs.
              0
            </div>
            <div>
              &gt;
            </div>
          </div>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
